### PR TITLE
[action] artifactory - add option to authenticate using an API key

### PR DESCRIPTION
**Pull request by fstaine**
Signed-off-by: STAINE Florian <florian.staine@orange.com>

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass (except the ones that where already not passing on master, not related to the Artifactory action)
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

The _Artifactory_ plugin allows to upload artifacts to a given Artifactory repository.
The lib used to send the artifacts allow to use a username/password or tu use an API key (prefered), but the lane was only accepting the username/password parameters. 

The objective of this PR to to enable logging in with an API key.

### Description

This changes allow to pass an 'api_key' parameter to the _artifactory_ lane action.
Verifications are made so that only api_key or username & password parameters are used, providing a clear message to the user if any other combination of crendentials is used.

### Testing Steps

To test this new feature, an Artifactory server is required.

To send an artifact using the API key:
`artifactory(api_key: <MY_API_KEY>, endpoint: <ENDPOINT>, file: <FILE.EXT>, repo: <REPO>, repo_path: <REPO_PATH>`

### CODE REVIEW

File: fastlane/swift/Fastlane.swift
Comment (rogerluan) [inline fastlane/swift/Fastlane.swift:700]: Code in all `.swift` files are auto-generated during the release process - could you revert these changes, please? 🙏

File: fastlane/lib/fastlane/actions/artifactory.rb
Comment (rogerluan) [inline fastlane/lib/fastlane/actions/artifactory.rb:129]:
```suggestion
                                         UI.user_error!("You can't use option '#{value.key}' along with 'username'")
```

File: fastlane/lib/fastlane/actions/artifactory.rb
Comment (rogerluan) [inline fastlane/lib/fastlane/actions/artifactory.rb:138]:
```suggestion
                                         UI.user_error!("You can't use option '#{value.key}' along with 'password'")
```

File: fastlane/lib/fastlane/actions/artifactory.rb
Comment (rogerluan) [inline fastlane/lib/fastlane/actions/artifactory.rb:147]:
```suggestion
                                        UI.user_error!("You can't use option '#{value.key}' along with 'api_key'")
```

File: fastlane/lib/fastlane/actions/artifactory.rb
Comment (rogerluan) [inline fastlane/lib/fastlane/actions/artifactory.rb:46]: I think it's more readable to validate these inputs on the top of this function using ruby one-liners, instead of adding more indentation with a `if/else` statement 🙇 
It'd look something like this:

```
UI.user_error!("Cannot connect to Artifactory - 'username' was provided but it's missing 'password'") if params[:username] && !params[:password]
UI.user_error!("Cannot connect to Artifactory - 'password' was provided but it's missing 'username'") if !params[:username] && params[:password]
UI.user_error!("Cannot connect to Artifactory - either 'api_key', or 'username' and 'password' must be provided") if !params[:api_key] && !params[:username]

if File.exist?(file_path) 
  …
else
  …
end
```

(also notice that I improved the copy of the error messages 😃

**Review by rogerluan — CHANGES_REQUESTED** [review]
Hey @fstaine 👋 

Welcome to _fastlane_ and thank you for your first contribution! 🎉 

I left some light comments on copy and code standardization, but your overall logic looks solid to me! 💪

**Comment by fstaine** [comment]
@rogerluan Changes applied, thank's for the review !

File: fastlane/lib/fastlane/actions/artifactory.rb
Comment (rogerluan) [inline fastlane/lib/fastlane/actions/artifactory.rb:18]: Can you run `bundle exec rubocop -a` again? I think this is causing CI to fail (trailing whitespace infraction)

**Review by rogerluan — CHANGES_REQUESTED** [review]
Thanks for addressing the changes! 💪 

Could you take a look at the failing tests? 🙏

**Comment by fstaine** [comment]
@rogerluan I updated the change but the I don't get error in 'Validate Fastlane.swift Generation', I'll let you take a look at it and tell me what to do ! 👮‍♂️

**Comment by rogerluan** [comment]
> @rogerluan I updated the change but I don't get the error in 'Validate Fastlane.swift Generation', I'll let you take a look at it and tell me what to do ! 👮‍♂️

I don't know either, we'll need @joshdholtz to take a look at this when he has the chance 😬 

Thanks for taking care of the other issues! 🙏

**Review by rogerluan — APPROVED** [review]

**Review by joshdholtz — CHANGES_REQUESTED** [review]
@fstaine This looks great! There are some failing tests that need fixing and I think it would be good to add some tests that reflect these changes.

I'm not able to add commits to this since it looks like this was forked to an organization and not your own GitHub account 🤔 If you want to give me write access to your fork I'd be happy to make the fixes to the tests! 💪 ❤️

**Comment by fstaine** [comment]
@joshdholtz 
I saw the failing tests but they are already failing on master and I don't know how to fix them 😢 
Unfortunately, I don't think that I can give you the access to the fork is it's owned by my company...
Maybe I can wait for the fixes to be merged onto master and then I'll rebase my work.

For the tests on my work, there isn't mutch to test as the work is mainly done by the `artifactory` gem, but I can add tests on the parameter verification !

**Comment by rogerluan** [comment]
@fstaine just a heads up that I believe the tests that are failing on your branch should be fixed if you rebase/merge master into your branch 😊 hope this helps.

And adding tests to the verify the parameters would be 💯 🤗

**Comment by fstaine** [comment]
Oh, indeed, I tought it wasn't fixed as I still had issues when running `rspec` in local !

I tried to add some tests, but I don't know how it works with required plugins...

I have the following error when running the test:
> cannot load such file -- artifactory

Due to the `require 'artifactory'` in the action...
Do you have any idea on how to fix this ? This doesn't seems to be an issue with some other `require ...` so I don't know how to fix this... 😢

**Comment by rogerluan** [comment]
Hm, not from the top of my head. May be an issue with bundler(?) 
As a workaround maybe you could install artifactory gem on your system (instead of only for this project). Not recommended, but might work?

**Comment by fstaine** [comment]
@rogerluan If I install the gem locally, the tests won't pass on CI & other machines so they will be useless
Sorry i'm not very experienced in Ruby 🙃

**Comment by fstaine** [comment]
OK, I think that I just need to add the 'artifacory' plugin into `fastlane.gemspec` !
I'm adding the tests !  🎉

**Review by joshdholtz — APPROVED** [review]
This looks great! Thank you so much for making this change 😊 Really appreciate the contribution ❤️

**Pull request merged**
